query-service-impl/src/test/java/org/hypertrace/core/query/service/pinot/ExecutionContextTest.java

```diff
@@ -308,4 +315,147 @@ public void testSelectionsLinkedHashSet() {
+    Builder builder = QueryRequest.newBuilder();
+    builder.addGroupBy(
+        Expression.newBuilder()
+            .setFunction(createTimeColumnGroupByFunction("SERVICE.startTime", "15000:MILLISECONDS"))
+            .build());
+    QueryRequest queryRequest = builder.build();
+    ExecutionContext context = new ExecutionContext("test", queryRequest);
+    assertEquals(Optional.of(Duration.ofSeconds(15)), context.getTimeSeriesPeriod());
+  }
+
+  @Test
+  public void testComputeTimeRangeDuration() {
+    QueryRequest queryRequest = getQueryRequestWithTimeFilter(Duration.ofMinutes(60));
+    ExecutionContext context = new ExecutionContext("test", queryRequest);
+    context.setTimeFilterColumn("SERVICE.startTime");
+    context.computeTimeRangeDuration(queryRequest);
+    assertEquals(Optional.of(Duration.ofMinutes(60)), context.getTimeRangeDuration());
+  }
+
+  private static QueryRequest getQueryRequestWithTimeFilter(Duration timeRange) {
+    long startTimeInMillis =
+        Instant.ofEpochSecond(TimeUnit.SECONDS.convert(1, TimeUnit.HOURS)).toEpochMilli();
+    Builder builder = QueryRequest.newBuilder();
+    builder.addGroupBy(
+        Expression.newBuilder()
+            .setFunction(createTimeColumnGroupByFunction("SERVICE.startTime", "15:SECONDS"))
+            .build());
+    Filter filter1 =
+        createFilter(
+            "SERVICE.startTime",
+            Operator.GE,
+            ValueType.LONG,
+            startTimeInMillis - timeRange.toMillis());
+    Filter filter2 =
+        createFilter("SERVICE.startTime", Operator.LT, ValueType.LONG, startTimeInMillis);
+    Filter filter3 = createFilter("SERVICE.id", Operator.NEQ, ValueType.NULL_STRING, "");
+
+    builder.setFilter(
+        Filter.newBuilder()
+            .setOperator(Operator.AND)
+            .addChildFilter(filter1)
+            .addChildFilter(filter2)
+            .addChildFilter(filter3)
+            .build());
+    return builder.build();
+  }
+
+  private static Function.Builder createTimeColumnGroupByFunction(
+      String timeColumn, String periodSecs) {
+    return Function.newBuilder()
+        .setFunctionName("dateTimeConvert")
+        .addArguments(createColumnExpression(timeColumn))
+        .addArguments(
+            Expression.newBuilder()
+                .setLiteral(
+                    LiteralConstant.newBuilder()
+                        .setValue(Value.newBuilder().setString("1:MILLISECONDS:EPOCH"))))
+        .addArguments(
+            Expression.newBuilder()
+                .setLiteral(
+                    LiteralConstant.newBuilder()
+                        .setValue(Value.newBuilder().setString("1:MILLISECONDS:EPOCH"))))
+        .addArguments(
+            Expression.newBuilder()
+                .setLiteral(
+                    LiteralConstant.newBuilder()
+                        .setValue(Value.newBuilder().setString(periodSecs))));
+  }
+
+  private static Expression.Builder createColumnExpression(String columnName) {
```

Comment:
So this one exists in the regular utils, but if it didn't can be added to the test utils file.